### ISSUE TITLE
Question about Classic Configuration Migration

### ISSUE THREAD

**Issue by austinarbor** [Hakky54/sslcontext-kickstart#434]
Hi @Hakky54, thanks for creating and maintaining this library, it looks really useful. I apologize in advance for the dumb question, but I just wanted to confirm my understanding was correct.

My use case is I want to use the default `cacerts` file, add some additional certificates to it at run time, and then use the combined list of certificates globally within the application. I _think_ this would be the equivalent of having a predefined truststore with the certificates already combined and passing it in by `-Djavax.net.ssl.trustStore=/path/to/my/cacerts -Djavax.net.ssl.trustStorePassword=changeit`, is that right?

Based on the code samples from your README, I think all I need to do is combine the examples:

```java
// build the default factory
SSLFactory sslFactory = SSLFactory.builder()
        .withDefaultTrustMaterial()
        .withSystemTrustMaterial()
        .withInflatableTrustMaterial()
        .build();

List<X509Certificate> certificates = ... ;
// dynamically add the certificates I want to 
TrustManagerUtils.addCertificate(sslFactory.getTrustManager().get(), certificates);

// this sets the trustStore as if I had passed it in -Djavax.net.ssl.trustStore=/path/to/my/cacerts ?
SSLContext.setDefault(sslFactory.getSslContext());
```

Is my understanding correct that the above would have the same effect as if all of the certificates were present in the default `cacerts` file from the beginning?

Do I need to worry about calling `KeyManagerUtils.swapKeyManager(..)` or `TrustManagerUtils.swapTrustManager(..)`?

Thank you in advance

**Comment by Hakky54** [comment]
Hello hello, thank you for your kind words ❤️ 

If you use the well known VM arguments for ssl, like `-Djavax.net.ssl.trustStore=/path/to/my/cacerts -Djavax.net.ssl.trustStorePassword=changeit` it will be used to construct the default SSLContext which will be used mostly globally. So if you don't want to use that VM argument and still want to have the content of cacerts, you should use: `.withDefaultTrustMaterial()` which just loads the content of cacerts.

I noticed you also use `.withSystemTrustMaterial()` which is fine, but not related to cacerts. It gets the operating system trusted ca's.

I tried to replicate your use case locally and it looks like this:

```java
import nl.altindag.ssl.SSLFactory;
import nl.altindag.ssl.util.KeyStoreUtils;
import nl.altindag.ssl.util.TrustManagerUtils;

import javax.net.ssl.X509ExtendedTrustManager;
import java.security.KeyStore;
import java.security.cert.X509Certificate;
import java.util.Arrays;

public class App {

    public static void main(String[] args) {
        SSLFactory sslFactory = SSLFactory.builder()
                .withDefaultTrustMaterial()
                .withInflatableTrustMaterial()
                .withSystemTrustMaterial()
                .withSwappableTrustMaterial()
                .build();

        System.out.println(sslFactory.getTrustedCertificates().size());

        KeyStore keyStore = KeyStoreUtils.loadKeyStore("keystore/truststore.jks", "secret".toCharArray());
        X509ExtendedTrustManager trustManager = TrustManagerUtils.createTrustManager(keyStore);
        X509Certificate[] newTrustedCertificates = trustManager.getAcceptedIssuers();

        TrustManagerUtils.addCertificate(sslFactory.getTrustManager().get(), Arrays.asList(newTrustedCertificates));
        System.out.println(sslFactory.getTrustedCertificates().size());

        TrustManagerUtils.swapTrustManager(sslFactory.getTrustManager().get(), trustManager);
        System.out.println(sslFactory.getTrustedCertificates().size());
    }

}
```

The output is: 
```
255
256
1
```

So if you use swapTrustManager or swapKeyManager it will mostly replace the in memory trustmanagers and keymanagers as it is designed for that kind of purpose. I noticed a small bug, currently it cannot handle adding a certificate when multiple trustmanagers have been used. So I will fix that soon.

So yes, you need to worry calling the swap methods as that will remove your earlier used keymanager/trustmanager. So either you need to reinitialize that setup and combine it with a new trustmanager/keymanager so you can provide that bundle to the swap methods. I think reinitializing would be more cleaner as your code will be also more readable. Also I would advice to provide a keystore path to the `withInflatableTrustMaterial` method as it will save the state, or else it will be lost on every startup of the app.

I hope it makes a bit sense, feel free to ask new questions

**Comment by austinarbor** [comment]
@Hakky54 thank you for your detailed response! Before finding this library I had written a lot of code to programmatically update the cacerts file and persist it...but then I realized I had no way of reloading the system TrustManager with my new certs! This library will actually replace all the code I wrote and then more!

I noticed you didn't call `SSLContext.setDefault(sslFactory.getSslContext());` is that not needed?

I apologize if this is a dumb question - I admit that this is my first time diving this deep into SSL with java

**Comment by Hakky54** [comment]
Glad to hear that you can replace your code with this library 😄 Really appreciate that it makes your use case much more easy to tackle.

> I noticed you didn't call SSLContext.setDefault(sslFactory.getSslContext()); is that not needed?

That is still needed if you want to configure your SSLContext globally. I just forgot to add that in the example code. By the way, way do you need to set it globally, can't you configure your server or client directly with the sslcontext?

**Comment by austinarbor** [comment]
I need to set it globally - the apps are short-lived (invoked via cli) but need to have certs conditionally loaded at runtime without restarting the CLI itself. It's a long story and agreed it's not best practice, but due to short lived nature of the app code running, "polluting" the context is not a concern

**Comment by Hakky54** [comment]
Ah, that makes more sense.

By the way, the code snippet which I shared, is not working yet as I discovered a bug today. The fix is here: https://github.com/Hakky54/sslcontext-kickstart/pull/436

I will resolve and make a release soon. Do you have any other doubts or do I need to help you to clarify ano other open topics? Just also checking whether this github issue can be closed or you have other issues.

**Comment by austinarbor** [comment]
I think you've answered all of my questions, I really appreciate the detailed and quick responses. Thanks again!

**Comment by Hakky54** [comment]
Hi @austinarbor I just made a release today, using version `8.3.0` should work correctly with your setup now as discussed here: https://github.com/Hakky54/sslcontext-kickstart/issues/434#issuecomment-1889318011. So Using inflatable trustmaterial, system trustmaterial, jdk trustmaterial, and custom keystores/certificates alongside with a swappable trustmanager

**Comment by austinarbor** [comment]
@Hakky54 thanks for the update!

**Comment by austinarbor** [comment]
@Hakky54 do you know if it is possible to use our SSLContext as the default TLS SSLContext as well? 

```java
@Test
  void sslContextTest() throws Exception {
    var factory = SSLFactory.builder().withDefaultTrustMaterial().withSystemTrustMaterial().build();

    var ctx = factory.getSslContext();
    SSLContext.setDefault(ctx);

    /*
    I tried this, but it had no effect
    var provider = ctx.getProvider();
    Security.insertProviderAt(provider, 1);
     */

    var context = SSLContext.getInstance("TLS");
    // fails, does not return our context
    assertThat(context).isSameAs(ctx);
  }
```

**Comment by Hakky54** [comment]
Well with the current version it is not possible, however with a small adjustment I can enable that kind of configuration. Although it is not complicated to enable it, I am not sure why it can be useful. You have the SSLFactory and the SSLContext already, why would you like to configure it like that? Can you maybe explain your use case, your problem statement so I am able to understand why this can be useful.

**Comment by austinarbor** [comment]
I inherited more than 30 repos of plugins. Currently our larger service relies on the system installed JRE, however, we are moving to a packaged JRE implementation where our own JRE comes with the software and we will no longer use the system's installed JRE. Unfortunately, _none_ of the repos did any SSLContext customization, they all relied on system defaults with the expectation of any required certificates being present in `cacerts`

Therefore, before, users were expected to install any custom certs into the JRE's `cacerts` file. Going forward that will not be possible.

All of these plugins rely on a common shared library. My _goal_, is to use your ssl-kickstart library to globally configure all of the SSLContexts from the plugin entry point to behave _as if_ any custom certificates were included in cacerts. Does that make sense?

As I mentioned before, due to the nature of how the plugins execute, I am not worried about best practices or polluting the context. Due to the nature of the codebase, I don't think it's feasible for me to audit everywhere an ssl context gets used and where I would need to pass one in. I am also concerned about future maintenance of the plugins - it is hard in PRs to remember to check (especially if I am not the one reviewing) - does this code path make secure http calls? does it need a custom ssl context?

The latest example is I found one of the plugins is using the jersey client. Looking through that code, if you don't provide your own ssl context it gets its default one from SSLContext.getInstance("TLS"). I'd like to be able to just globally set our custom context as the one that gets returned for cases like this.

**Comment by Hakky54** [comment]
Now it makes more sense. Thank you for giving more context. I checked even the source code of the Jersey client and indeed they just did what you mentioned, which is odd in my opinion, but maybe they had a good reason for it. It is also mentioned in their javadoc https://javadoc.io/static/com.sun.jersey/jersey-client/1.19/index.html?com/sun/jersey/client/urlconnection/HTTPSProperties.html

I decided to give it a shot. I created a new branch with the code changes, see here: https://github.com/Hakky54/sslcontext-kickstart/pull/454

You can fork and build a local artifact to try it out on your side. A basic usage would be something like this:
```java
import nl.altindag.ssl.provider.FenixProvider;
import nl.altindag.ssl.provider.SSLFactoryProvider;

import javax.net.ssl.SSLContext;
import java.security.NoSuchAlgorithmException;
import java.security.Security;

public class App {

    public static void main(String[] args) throws NoSuchAlgorithmException {
        FenixProvider provider = new FenixProvider();
        Security.insertProviderAt(provider, 1);

        SSLFactory sslFactory = SSLFactory.builder()
                .withDefaultTrustMaterial()
                .withSystemTrustMaterial()
                .build();

        SSLFactoryProvider.set(sslFactory);

        SSLContext sslContext = SSLContext.getInstance("TLS");
    }

}
```

With this code snippet you can create a FenixProvider and give it to the Security utility of java. The FenixProvider knows it should use a custom SSLContext and it will instruct java to do so. So when you call `SSLContext.getInstance("TLS")` it will load the sslcontext from the sslfactory. Next to that you need to construct your base SSLFactory and give that to the SSLFactoryProvider. 

I have added a screenshot below of my debugger and as you can see the newly created SSLContext with the value TLS is just using the one which is configured earlier.

<img width="2048" alt="Screenshot 2024-02-09 at 23 33 43" src="https://github.com/Hakky54/sslcontext-kickstart/assets/16032204/5a802a66-6d02-451d-b0cd-135aa5ff66d2">

Can you for this project, checkout this branch and built a local artifact on your side? In that way you can try whether it will also work on your side. Looking forward to your test results.

**Comment by austinarbor** [comment]
@Hakky54 wow that worked perfectly!

**Comment by Hakky54** [comment]
Awesome, great to hear that it works. I will add some tests and maybe I will rename some classes.  I will finalize it in couple of days. If anything needs to be adjusted feel free to provide feedback in the pull request or else I will keep you posted when it is finished

**Comment by austinarbor** [comment]
Thanks again! Really appreciate it

**Comment by Hakky54** [comment]
Hi @austinarbor I simplified the usage, hoe can still find the examples here: https://github.com/Hakky54/sslcontext-kickstart?tab=readme-ov-file#global-ssl-configuration

I also made a release today, so it is available in version 8.3.2
